design/downstream-crl-design.md

```diff
@@ -0,0 +1,128 @@
+
+## Goals
+- Allow a Kubernetes secret containing a PEM encoded CRL to be used when validating client connections
+- Perform simple validation of the K8s secret
+  - Validate that it is indeed a PEM CRL object
+
+## Non Goals
+- Upstream Validation
+- Support for `only_verify_leaf_cert_crl`
+  - Not in a released version of envoy yet, still in development
+
+
+## High-Level Design
+The design would for the most part mirror the implementation of `DownstreamValidation.CACertificate` and `DownstreamValidation.SkipClientCertValidation`.
+
+Sample YAML
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: client-validation-example
+  namespace: default
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: server-credentials
+      clientValidation:
+        caSecret: ca-cert-for-client-validation
+        caCrlSecret: crl-for-client-validation
+  routes:
+    - services:
+        - name: service
+          port: 80
+```
+
+## Detailed Design
+
+### CRL Secret
+
+The same approach shall be followed for configuring revocation lists as is used currently to store the CA certificates for client validation
+
+The CRL is stored in an opaque Kubernetes secret.
+The secret will be stored in the same namespace as the corresponding `HTTPProxy` object.
+The secret object shall contain entry named `crl.pem`.
+The constents shall be the CRL in PEM format.
+The file may contain "PEM bundle", that is, a list of CRLs concatenated in single file.
+
+Example:
+
+```bash
+kubectl create secret generic crl-for-client-validation --from-file=./crl.pem
+```
+
+### httpproxy.DownstreamValidation additions
+
+```go
+// DownstreamValidation defines how to verify the client certificate.
+type DownstreamValidation struct {
+  ...
+
+	// Name of a Kubernetes Opaque secret that contains a concatenated list of
+  // pem encoded crls.
+	// This will be used to verify that a client certificate has not been revoked
+	// +optional
+	// +kubebuilder:validation:MinLength=1
+	CACertificateRevocationList string `json:"caCrlSecret,omitempty"`
+}
+```
+
+### Envoy Configuration
+
+The new fields from `spec.virtualhost.tls.clientValidation` must be parsed and mapped to `auth.CommonTlsContext.ValidationContextType`
+
+- `spec.virtualhost.tls.clientValidation.caCrlSecret` -> `envoy_v3_tls.CommonTlsContext_ValidationContext.Crl`
+  - `internal.envoy.v3.auth.validationContext()` will be updated to accept a `crl` the same way it accepts a `ca`
+  - If empty it will not set the `crl` field, maintaining backward compatibility
+
+### Secret validation
+
+Currently basic validation is done on Opaque secrets with a `ca.crt` key, just to make sure that the length is non-zero.
+This validation would also be performed for the CRL.
+
+Validation is also performed to make sure that the CA Bundle has the correct PEM header and is of type `CERTIFICATE`.
+The same validation would be performed, but checking for type `CRL`
```

Comment:
I think the type is `X509 CRL` https://datatracker.ietf.org/doc/html/rfc7468#section-4